python/src/scanner.rs

```diff
@@ -80,25 +84,29 @@ impl Scanner {
     }
 
     #[staticmethod]
-    fn from_fragments(fragments: Vec<FileFragment>) -> PyResult<Self> {
-        let fragments = fragments
-            .into_iter()
-            .map(|f| f.fragment().clone())
-            .collect::<Vec<_>>();
-
-        if fragments.is_empty() {
-            return Err(PyValueError::new_err("No fragments provided"));
-        }
-
-        // TODO: implement conversion traits so we can avoid the clones.
-
-        let dataset = Arc::new(fragments[0].dataset().clone());
-        let fragments = fragments
-            .into_iter()
-            .map(|f| f.metadata().clone())
-            .collect::<Vec<_>>();
+    fn from_fragments(fragments: PyObject, dataset: Dataset) -> PyResult<Self> {
+        let fragment_iter = std::iter::from_fn(move || {
+            Python::with_gil(|py| match fragments.call_method0(py, "__next__") {
+                Ok(fragment) => {
+                    let fragment = fragment.extract::<FileFragment>(py).unwrap();
+                    let fragment: LanceFragment = fragment.into();
+                    Some(Ok(FormatFragment::from(fragment)))
+                }
+                Err(err) => {
+                    if err.is_instance_of::<PyStopIteration>(py) {
+                        None
+                    } else {
+                        Some(Err(PyValueError::new_err(format!(
+                            "Failed to get fragment: {}",
+                            err
+                        ))))
+                    }
+                }
+            })
+        });
 
-        let scanner = LanceScanner::from_fragments(dataset, fragments);
+        let fragments: Vec<FormatFragment> = fragment_iter.collect::<PyResult<_>>()?;
```

Comment:
Is the logic here to collect a list / vector of fragments in rust instead in python? 

are we going to make Fragments to be an iterator  in rust later? Some of users expected like 500k+ fragments .

Reply:
Yes, this is just set up for later. Once we refactor `LanceScanner` to be able to take an iterator, we will remove this line. I didn't feel like I was ready to do that refactor yet, as I think it will change quite a few lines of code that I don't yet have a strong understanding of.